File: church/report/parish_dashboard/parish_dashboard.py
import frappe
from frappe import _
from datetime import datetime

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    chart = get_chart_data(data)
    summary = get_summary(data)
    return columns, data, None, chart, summary

def get_columns():
    return [
        {
            "fieldname": "name",
            "label": _("Name"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "type",
            "label": _("Type"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "total_members",
            "label": _("Total Members"),
            "fieldtype": "Int",
            "width": 120
        },
        {
            "fieldname": "total_contributions",
            "label": _("Total Contributions"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "avg_contribution_per_member",
            "label": _("Avg Contribution/Member"),
            "fieldtype": "Currency",
            "width": 180
        },
        {
            "fieldname": "last_contribution_date",
            "label": _("Last Contribution Date"),
            "fieldtype": "Date",
            "width": 150
        }
    ]

def get_data(filters):
    data = []
    parishes = get_parishes(filters)
    
    for parish in parishes:
        parish_data = get_entity_data("Parish", parish.name, filters)
        parish_data["parent"] = None
        data.append(parish_data)
        
        sub_parishes = get_sub_parishes(parish.name)
        for sub_parish in sub_parishes:
            sub_parish_data = get_entity_data("Sub Parish", sub_parish.name, filters)
            sub_parish_data["parent"] = parish.name
            data.append(sub_parish_data)

    return data

def get_entity_data(doctype, name, filters):
    total_members = get_total_members(doctype, name)
    total_contributions, last_contribution_date = get_contribution_data(doctype, name, filters)
    avg_contribution = total_contributions / total_members if total_members else 0
    
    return {
        "name": name,
        "type": doctype,
        "total_members": total_members,
        "total_contributions": total_contributions,
        "avg_contribution_per_member": avg_contribution,
        "last_contribution_date": last_contribution_date
    }

def get_parishes(filters):
    parish_filters = {}
    if filters.get("diocese"):
        parish_filters["diocese"] = filters.get("diocese")
    if filters.get("archdeaconry"):
        parish_filters["archdeaconry"] = filters.get("archdeaconry")
    if filters.get("parish"):
        parish_filters["name"] = filters.get("parish")
    return frappe.get_all("Parish", filters=parish_filters, fields=["name"])

def get_sub_parishes(parish):
    return frappe.get_all("Sub Parish", filters={"parish": parish}, fields=["name"])

def get_total_members(doctype, name):
    if doctype == "Parish":
        return frappe.db.count("Church Member", filters={"parish": name})
    elif doctype == "Sub Parish":
        return frappe.db.count("Church Member", filters={"sub_parish": name})
    return 0

def get_contribution_data(doctype, name, filters):
    date_filter = []
    if filters.get("from_date"):
        date_filter.append(["payment_date", ">=", filters.get("from_date")])
    if filters.get("to_date"):
        date_filter.append(["payment_date", "<=", filters.get("to_date")])

    if doctype == "Parish":
        members = frappe.get_all("Church Member", filters={"parish": name}, pluck="name")
    elif doctype == "Sub Parish":
        members = frappe.get_all("Church Member", filters={"sub_parish": name}, pluck="name")
    else:
        return 0, None

    if not members:
        return 0, None

    result = frappe.db.sql("""
        SELECT SUM(amount) as total, MAX(payment_date) as last_date
        FROM `tabChurch Payment`
        WHERE member IN (%s) AND docstatus = 1
        %s
    """ % (', '.join(['%s'] * len(members)), 
           "AND " + " AND ".join([f"{f[0]} {f[1]} %s" for f in date_filter]) if date_filter else ""),
    tuple(members + [f[2] for f in date_filter]), as_dict=True)

    total = result[0].total if result and result[0].total else 0
    last_date = result[0].last_date if result and result[0].last_date else None

    return total, last_date

def get_chart_data(data):
    labels = [d['name'] for d in data if d['type'] == 'Parish']
    contributions = [d['total_contributions'] for d in data if d['type'] == 'Parish']
    members = [d['total_members'] for d in data if d['type'] == 'Parish']

    return {
        "data": {
            "labels": labels,
            "datasets": [
                {
                    "name": "Total Contributions",
                    "values": contributions
                },
                {
                    "name": "Total Members",
                    "values": members
                }
            ]
        },
        "type": "bar",
        "colors": ["#4CAF50", "#2196F3"]
    }

def get_summary(data):
    parish_data = [d for d in data if d['type'] == 'Parish']
    total_members = sum(d['total_members'] for d in parish_data)
    total_contributions = sum(d['total_contributions'] for d in parish_data)
    avg_contribution = total_contributions / total_members if total_members else 0
    last_contribution_date = max([d['last_contribution_date'] for d in parish_data if d['last_contribution_date']], default=None)

    return [
        {
            "value": total_members,
            "label": _("Total Members"),
            "datatype": "Int",
            "currency": None
        },
        {
            "value": total_contributions,
            "label": _("Total Contributions"),
            "datatype": "Currency",
            "currency": frappe.defaults.get_global_default("currency")
        },
        {
            "value": avg_contribution,
            "label": _("Average Contribution per Member"),
            "datatype": "Currency",
            "currency": frappe.defaults.get_global_default("currency")
        },
        {
            "value": last_contribution_date,
            "label": _("Last Contribution Date"),
            "datatype": "Date",
            "currency": None
        }
    ]

File: church/report/parish_dashboard/parish_dashboard.js
frappe.query_reports["Parish Dashboard"] = {
    "filters": [
        {
            "fieldname": "diocese",
            "label": __("Diocese"),
            "fieldtype": "Link",
            "options": "Diocese",
            "get_query": function() {
                return {
                    filters: {
                        "is_group": 1
                    }
                };
            }
        },
        {
            "fieldname": "archdeaconry",
            "label": __("Archdeaconry"),
            "fieldtype": "Link",
            "options": "Archdeaconry",
            "get_query": function() {
                var diocese = frappe.query_report.get_filter_value('diocese');
                return {
                    filters: {
                        "diocese": diocese
                    }
                };
            },
            "depends_on": "diocese"
        },
        {
            "fieldname": "parish",
            "label": __("Parish"),
            "fieldtype": "Link",
            "options": "Parish",
            "get_query": function() {
                var archdeaconry = frappe.query_report.get_filter_value('archdeaconry');
                return {
                    filters: {
                        "archdeaconry": archdeaconry
                    }
                };
            },
            "depends_on": "archdeaconry"
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        }
    ],
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        if (column.fieldname == "total_contributions" && data && data.total_contributions) {
            value = "<span style='color:green'>" + value + "</span>";
        }
        if (column.fieldname == "avg_contribution_per_member" && data && data.avg_contribution_per_member) {
            value = "<span style='color:blue'>" + value + "</span>";
        }
        return value;
    },
    "tree": true,
    "name_field": "name",
    "parent_field": "parent",
    "initial_depth": 1,
    "onload": function(report) {
        report.page.add_inner_button(__("View on Map"), function() {
            frappe.set_route("gps-dashboard", {"entity_type": "Parish"});
        });
    },
    "get_datatable_options": function(options) {
        return Object.assign(options, {
            treeView: true,
            cellHeight: 45,
            columns: [
                { name: "name", width: 300 },
                { name: "type", width: 100 },
                { name: "total_members", width: 120 },
                { name: "total_contributions", width: 150 },
                { name: "avg_contribution_per_member", width: 180 },
                { name: "last_contribution_date", width: 150 }
            ]
        });
    },
    "get_chart_data": function(columns, result) {
        return {
            data: {
                labels: result.filter(d => d.type === "Parish").map(d => d.name),
                datasets: [
                    {
                        name: "Total Members",
                        values: result.filter(d => d.type === "Parish").map(d => d.total_members)
                    },
                    {
                        name: "Total Contributions",
                        values: result.filter(d => d.type === "Parish").map(d => d.total_contributions)
                    }
                ]
            },
            type: 'bar',
            colors: ['#7cd6fd', '#743ee2']
        }
    },
    "after_datatable_render": function(datatable_obj) {
        $(datatable_obj.wrapper).find(".dt-row-0").find('input[type=checkbox]').click();
    },
    "get_chart_options": function(columns, result) {
        return {
            title: {
                text: 'Parish Overview'
            },
            subtitle: {
                text: 'Members and Contributions'
            },
            xAxis: {
                title: {
                    text: 'Parish'
                }
            },
            yAxis: [
                {
                    title: {
                        text: 'Total Members'
                    }
                },
                {
                    title: {
                        text: 'Total Contributions'
                    },
                    opposite: true
                }
            ]
        }
    }
};